tests/test_frompyobject.rs

```diff
@@ -0,0 +1,308 @@
+        a: String,
+    },
+    #[transparent]
+    CatchAll(&'a PyAny),
+}
+
+#[pyclass]
+pub struct PyBool {
+    #[pyo3(get)]
+    bla: bool,
+}
+
+#[test]
+fn test_enum() {
+    let gil = Python::acquire_gil();
+    let py = gil.python();
+
+    let tup = PyTuple::new(py, &[1.into_py(py), "test".into_py(py)]);
+    let f = Foo::extract(tup.as_ref()).expect("Failed to extract Foo from tuple");
+    match f {
+        Foo::TupleVar(test, test2) => {
+            assert_eq!(test, 1);
+            assert_eq!(test2, "test");
+        }
+        _ => panic!("Expected extracting Foo::TupleVar, got {:?}", f),
+    }
+
+    let pye = PyE {
+        test: "foo".into(),
+        test2: 0,
+    }
+    .into_py(py);
+    let f = Foo::extract(pye.as_ref(py)).expect("Failed to extract Foo from PyE");
+    match f {
+        Foo::StructVar { test } => assert_eq!(test.to_string_lossy(), "foo"),
+        _ => panic!("Expected extracting Foo::StructVar, got {:?}", f),
+    }
+
+    let int: PyObject = 1.into_py(py);
+    let f = Foo::extract(int.as_ref(py)).expect("Failed to extract Foo from int");
+    match f {
+        Foo::TransparentTuple(test) => assert_eq!(test, 1),
+        _ => panic!("Expected extracting Foo::TransparentTuple, got {:?}", f),
+    }
+    let none = py.None();
+    let f = Foo::extract(none.as_ref(py)).expect("Failed to extract Foo from int");
+    match f {
+        Foo::TransparentStructVar { a } => assert!(a.is_none()),
+        _ => panic!("Expected extracting Foo::TransparentStructVar, got {:?}", f),
+    }
+
+    let pybool = PyBool { bla: true }.into_py(py);
+    let f = Foo::extract(pybool.as_ref(py)).expect("Failed to extract Foo from PyBool");
+    match f {
+        Foo::StructVarGetAttrArg { a } => assert!(a),
+        _ => panic!("Expected extracting Foo::StructVarGetAttrArg, got {:?}", f),
+    }
+
+    let dict = PyDict::new(py);
+    dict.set_item("a", "test").expect("Failed to set item");
+    let f = Foo::extract(dict.as_ref()).expect("Failed to extract Foo from dict");
+    match f {
+        Foo::StructWithGetItem { a } => assert_eq!(a, "test"),
+        _ => panic!("Expected extracting Foo::StructWithGetItem, got {:?}", f),
+    }
+
+    let dict = PyDict::new(py);
+    dict.set_item("foo", "test").expect("Failed to set item");
+    let f = Foo::extract(dict.as_ref()).expect("Failed to extract Foo from dict");
+    match f {
+        Foo::StructWithGetItemArg { a } => assert_eq!(a, "test"),
+        _ => panic!("Expected extracting Foo::StructWithGetItemArg, got {:?}", f),
+    }
+
+    let dict = PyDict::new(py);
+    let f = Foo::extract(dict.as_ref()).expect("Failed to extract Foo from dict");
+    match f {
+        Foo::CatchAll(any) => {
+            let d = <&PyDict>::extract(any).expect("Expected pydict");
+            assert!(d.is_empty());
+        }
+        _ => panic!("Expected extracting Foo::CatchAll, got {:?}", f),
+    }
+}
+
+#[derive(FromPyObject)]
+pub enum Bar {
+    #[rename_err = "str"]
```

Comment:
Similarly I think annotation can be used more broadly later (if we started being able to generate type hints from these, for example).

```suggestion
    #[pyo3(annotation = "str")]
```